Clients/Xamarin.Interactive.Client.Web/ClientApp/Workbook.ts

```diff
@@ -0,0 +1,190 @@
+
+    if (!gistZipResponse.ok)
+        throw new Error("Couldn't download Gist from GitHub.");
+
+    var gistZip = await gistZipResponse.blob();
+    const workbookPackage = new File([ gistZip ], `${gistId}.zip`, {
+        type: "application/zip"
+    });
+    return await loadWorkbookFromWorkbookPackage(workbookSession, workbookPackage);
+}
+
+export async function loadWorkbookFromWorkbookPackage(workbookSession: WorkbookSession, workbookPackage: File): Promise<Workbook> {    
+    const loadedZip = await loadAsync(workbookPackage)
+    const workbookFiles = loadedZip.filter((path, file) => path.endsWith(".workbook"));
+
+    if (workbookFiles.length === 0)
+        throw new Error("No workbooks in workbook package.");
+
+    // Grab index.workbook or the first workbook file if there isn't an index.
+    const workbookFile = workbookFiles.find(wf => wf.name === "index.workbook") || workbookFiles[0];
+    const workbookFileContents = await workbookFile.async("text")
+
+    // TODO: Pass down _all_ of the files so that we can implement #r and #load.
+    return await loadWorkbookFromString(workbookSession, workbookPackage.name, workbookFileContents)
+}
+
+export interface WorkbookManifest {
+    packages: List<PackageDescription>
+    platforms: List<WorkbookTarget>
+    title?: string
+    id?: string
+    uti: string
+    rest: {}
+}
+
+export class Workbook {
+    private readonly _rawMetadata: any
+    private readonly _markdownContent: string
+    private readonly _manifest: WorkbookManifest
+    private readonly _fileName?: string
+
+    get rawMetadata(): any {
+        return this._rawMetadata
+    }
+
+    get markdownContent(): string {
+        return this._markdownContent
+    }
+
+    get manifest(): WorkbookManifest {
+        return this._manifest
+    }
+
+    get fileName(): string | undefined {
+        return this._fileName
+    }
+
+    constructor(workbookSession: WorkbookSession, fileName: string, markdownContent: string, metadata: any) {
+        this._fileName = fileName;
+        this._markdownContent = markdownContent.trim()
+        this._rawMetadata = metadata
+        this._manifest = this.parseManifest(this._rawMetadata, workbookSession)
+    }
+
+    getManifestToSave(): any {
+        const saveableManifest: any = Object.assign({
+            id: this.manifest.id,
+            uti: this.manifest.uti,
+            title: this.manifest.title,
+        }, this.manifest.rest)
+        saveableManifest.packages = this.manifest.packages.map(pd => {
+            return !pd ? undefined : { id: pd.packageId, version: pd.identityVersion }
+        }).filter((_: any) => _).toArray()
+        saveableManifest.platforms = this.manifest.platforms.map(wt => {
+            return !wt ? undefined : getPlatformFromWorkbookTargetId(wt.id)
+        }).filter((_: any) => _).toArray()
+        return saveableManifest
+    }
+
+    parseManifest(rawMetadata: any, workbookSession: WorkbookSession): WorkbookManifest {
```

Comment:
Would be good to explain the general idea of how this whole method works. It's a little hard to follow with all of the deleting properties out of `rawMetadata`. Should also explain that it will round-trip any and all arbitrary YAML and just pulls out bits that we actually need.